api/billingd/filrewards/filrewards.go

```diff
@@ -0,0 +1,217 @@
+)
+
+type Event int
+
+const (
+	Unspecified Event = iota
+	FirstKeyAccountCreated
+	FirstKeyUserCreated // does this fit "register first user?"
+	FirstOrgCreated
+	InitialBillingSetup
+	FirstBucketCreated
+	FirstBucketArchiveCreated
+	FirstMailboxCreated
+	FirstThreadDbCreated
+)
+
+const baseAttoFILReward = 1000 // What should this be? Should we read it from mongo?
+
+// maybe we want to read the meta values from mongo so we can update live.
+var eventRewards = map[Event]reward{
+	FirstKeyAccountCreated:    {factor: 3},
+	FirstKeyUserCreated:       {factor: 1},
+	FirstOrgCreated:           {factor: 3},
+	InitialBillingSetup:       {factor: 1},
+	FirstBucketCreated:        {factor: 2},
+	FirstBucketArchiveCreated: {factor: 2},
+	FirstMailboxCreated:       {factor: 1},
+	FirstThreadDbCreated:      {factor: 1},
+}
+
+type reward struct {
+	factor int
+}
+
+type RewardRecord struct {
+	Key               string          `bson:"key"`
+	AccountType       mdb.AccountType `bson:"account_type"`
+	Event             Event           `bson:"event"`
+	Factor            int             `bson:"factor"`
+	BaseAttoFILReward int             `bson:"base_atto_fil_reward"`
+	CreatedAt         time.Time       `bson:"created_at"`
+}
+
+type FilRewards struct {
+	col          *mongo.Collection
+	rewardsCache map[string]map[Event]struct{}
+	// ToDo: Need some service to send out notifications when rewards are met.
+}
+
+type Config struct {
+	DBURI          string
+	DBName         string
+	CollectionName string
+}
+
+func New(ctx context.Context, config Config) (*FilRewards, error) {
+	client, err := mongo.Connect(ctx, options.Client().ApplyURI(config.DBURI))
+	if err != nil {
+		return nil, fmt.Errorf("connecting to mongo: %v", err)
+	}
+	db := client.Database(config.DBName)
+	col := db.Collection(config.CollectionName)
+	_, err = col.Indexes().CreateMany(ctx, []mongo.IndexModel{
+		{
+			Keys:    bson.D{primitive.E{Key: "key", Value: 1}, primitive.E{Key: "event", Value: 1}},
+			Options: options.Index().SetUnique(true).SetSparse(true),
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "key", Value: 1}},
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "event", Value: 1}},
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "created_at", Value: 1}},
+		},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("creating indexes: %v", err)
+	}
+	// Populate cache.
+	opts := options.Find()
+	cursor, err := col.Find(ctx, bson.M{}, opts)
+	if err != nil {
+		return nil, fmt.Errorf("querying RewardRecords to populate cache: %s", err)
+	}
+	defer cursor.Close(ctx)
+	cache := make(map[string]map[Event]struct{})
+	for cursor.Next(ctx) {
+		var rec RewardRecord
+		if err := cursor.Decode(&rec); err != nil {
+			return nil, fmt.Errorf("decoding RewardRecord while building cache: %v", err)
+		}
+		ensureKeyEventCache(cache, rec.Key)
+		cache[rec.Key][rec.Event] = struct{}{}
+	}
+	if err := cursor.Err(); err != nil {
+		return nil, fmt.Errorf("iterating cursor while building cache: %v", err)
+	}
```

Comment:
We populate that rewards cache on startup. We'll want to keep an eye on how long this takes since it will grow literally with the amount of rewards stored over time.